### PR TITLE
Feature/backend/verify signature

### PR DESCRIPTION
**Pull request by gdchamal**
Apiv2 and apiv1 verify device signature. For the moment only log result. Will reject call later

### CODE REVIEW

File: src/backend/interfaces/REST/go.server/middlewares/authentication.go [hunk 1]
```diff
@@ -5,16 +5,61 @@
...
+}
+
+// getSignedQuery, build the HTTP query that has been signed
+func getSignedQuery(c * gin.Context) string {
+	query := c.Request.Method + c.Request.URL.String()
+	return query
+}
+
+// verifySignature, check for validity of device ecdsa signature
+func verifySignature(signature string, query string, curve string, x big.Int, y big.Int) (bool, error) {
```
Comment (sapiens-sapide) [inline src/backend/interfaces/REST/go.server/middlewares/authentication.go]: shorter func signature could be : `verifySignature(signature, query, curve string, x, y big.Int) (bool, error) {`

File: src/backend/interfaces/REST/go.server/middlewares/authentication.go [hunk 2]
```diff
@@ -5,16 +5,61 @@
 package http_middleware
 
 import (
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/sha256"
+	"encoding/asn1"
 	"encoding/base64"
+	"errors"
+	"log"
```
Comment (sapiens-sapide) [inline src/backend/interfaces/REST/go.server/middlewares/authentication.go]: we make use of `logrus` to log, not standard pkg `log`. Replace this import by  
 `log "github.com/Sirupsen/logrus"`

File: src/backend/interfaces/REST/go.server/middlewares/authentication.go [hunk 3]
```diff
@@ -48,6 +93,21 @@ func BasicAuthFromCache(cache backends.APICache, realm string) gin.HandlerFunc {
...
 		}
 
+		if device_sign, ok := c.Request.Header["X-Caliopen-Device-Signature"]; ok {
+			query := getSignedQuery(c)
+			valid, err := verifySignature(device_sign[0], query, auth.Curve, auth.X, auth.Y)
+			if err != nil {
+				log.Println("Error during signature verification: ", err)
+				kickUnauthorizedRequest(c, "Authorization error")
+			}
+			if valid == false {
```
Comment (sapiens-sapide) [inline src/backend/interfaces/REST/go.server/middlewares/authentication.go]: `if !valid`

File: src/backend/interfaces/REST/go.server/middlewares/authentication.go [hunk 4]
```diff
@@ -48,6 +93,23 @@ func BasicAuthFromCache(cache backends.APICache, realm string) gin.HandlerFunc {
...
+			valid, err := verifySignature(device_sign[0], query, auth.Curve, auth.X, auth.Y)
+			if err != nil {
+				log.Println("Error during signature verification: ", err)
+				// kickUnauthorizedRequest(c, "Authorization error")
+			}
+			if valid == false {
+				log.Println("Verification of signature failed")
+				// kickUnauthorizedRequest(c, "Authorization failed")
+			} else {
+				log.Println("Verofication of signature OK")
```
Comment (MonsieurLanza) [inline src/backend/interfaces/REST/go.server/middlewares/authentication.go]: small typo "Verofication".

File: src/backend/interfaces/REST/go.server/middlewares/authentication.go (hunk 2 above)
Comment (gdchamal) [reply src/backend/interfaces/REST/go.server/middlewares/authentication.go]: ok done

File: src/backend/interfaces/REST/go.server/middlewares/authentication.go (hunk 4 above)
Comment (gdchamal) [reply src/backend/interfaces/REST/go.server/middlewares/authentication.go]: fixed

File: src/backend/interfaces/REST/go.server/middlewares/authentication.go (hunk 1 above)
Comment (gdchamal) [reply src/backend/interfaces/REST/go.server/middlewares/authentication.go]: I wasn't aware of this shortcut. Thanks I rewrite using this syntax

**Pull request merged**